lucene/core/src/java/org/apache/lucene/index/EagerCheapMergePolicy.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.index;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+import org.apache.lucene.util.InfoStream;
+
+/**
+ * Eagerly proposes cheap merges on a commit. The goal is to help keep the segment count low where
+ * we can do so cheaply, particularly on a commit so that the IndexSearcher has fewer segments to
+ * search over.  This policy only overrides {@link #findFullFlushMerges(MergeTrigger, SegmentInfos, MergeContext)}; the
+ * rest are delegated to another policy.
+ */
+public class EagerCheapMergePolicy extends FilterMergePolicy {
```

Comment:
Could we rename this to `MergeOnCommitMergePolicy` perhaps?

Reply:
That's a nice vote of confidence I suppose because that's the generic name and it'd mean attempts from others to add a MP that merges on commit likewise would get the most obvious name.  For example "Highlighter" is forever people's go-to highlighter despite there being a better option.   If you like it so much, I do :-), maybe TieredMergePolicy might have this algorithm a setMaxCommitMergeWaitSeconds of 0 would mean it's never called any way and is thus "safe" to those who don't want it.  Just a thought and perhaps deserves JIRA level commentary instead of hidden in a GitHub PR code review.